src/compiler/scala/tools/nsc/transform/Constructors.scala

```diff
@@ -764,7 +764,7 @@ abstract class Constructors extends Statics with Transform with TypingTransforme
           primaryConstrBody.expr)
       })
 
-      if (omittableAccessor.exists(_.isOuterField) && !constructorStats.exists(_.exists { case i: Ident if i.symbol.isOuterParam => true; case _ => false}))
+      if ((exitingPickler(clazz.isAnonymousClass) || clazz.originalOwner.isTerm) && omittableAccessor.exists(_.isOuterField) && !constructorStats.exists(_.exists { case i: Ident if i.symbol.isOuterParam => true; case _ => false}))
```

Comment:
Agreed.

I initially tried to put the condition into `Delambdafy` such that it only applies the "pass null rather than outer" trick if the current owner of the call site of the constructor and the class being constructed share the same enclosing top level class.

Note that just sharing a compilation unit isn't enough as top level definitions in a file can be split into separate files.